### PR TITLE
*: refactor configuration

### PR DESCRIPTION
**Pull request by BusyJay**
In the past, we get a toml struct from the configuration file, and then deal with every configuration options again and again. This way will have following shortcomings:
1. Parsing logic spread around, which is hard to read and maintain;
2. Default values occur in several places, if they need to be changed some day, some places are easy to be missed.
3. Config parsing is mixed with rocksdb option construction, which is quite easy to make mistake.

This pr tries to make things clear by breaking the procedure into several phases.
1. Parse all configuration;
2. Adjust configuration base on command line args;
3. Construct services and struct according to the configuration.

And serde is used here to ease the parsing job, hence we don't need to write all the basic helper function again.

But this pr also introduces several breaking changes:
- *log-level*, *log-file* are moved from server section to the root.
- *pd.endpoints* is changed to array, and *https://* prefix is not allowed anymore.
- *server.capacity* is renamed as *raftstore.capacity*.
- *server.labels* is changed to an array.
- *server.store* is deleted and *server.data-dir* is renamed as *storage.data-dir*.
- *server.backup* is removed, and *server.backup_dir* is renamed as *rocksdb.backup_dir*.
- Kb, mb, gb are not supported anymore, upper case is required.
- error is reported to stderr before configuration is created and validated.

### CODE REVIEW

File: src/bin/tikv-server.rs [hunk 1]
```diff
@@ -45,882 +47,137 @@ mod signal_handler;
...
+        exit_with_msg(format!("{:?}", e));
     }
 
     for e in util::config::check_kernel() {
         warn!("{:?}", e);
     }
 
-    if !cfg!(windows) && env::var("TZ").is_err() {
+    if cfg!(unix) && env::var("TZ").is_err() {
         env::set_var("TZ", "/etc/localtime");
```
Comment (overvenus) [inline src/bin/tikv-server.rs]: `TZ` should be somethink like `Asia/Shanghai`, not a path.

File: src/bin/tikv-server.rs [hunk 2]
```diff
@@ -1023,108 +331,61 @@ fn main() {
...
-
-    let mut cfg = build_cfg(&matches, &config, cluster_id, addr, total_cpu_num as usize);
-    cfg.labels = get_store_labels(&matches, &config);
-    let (store_path, backup_path) = get_data_and_backup_dirs(&matches, &config);
-    cfg.storage.data_dir = store_path;
-
+        .unwrap_or_else(|e| exit_with_err(e));
     if cluster_id == DEFAULT_CLUSTER_ID {
-        panic!("in raftkv, cluster_id must greater than 0");
+        exit_with_msg(format!("cluster id can't be {}", cluster_id));
```
Comment (andelf) [inline src/bin/tikv-server.rs]: since it will always be DEFAULT_CLUSTER_ID

File: src/config.rs [hunk 3]
```diff
@@ -0,0 +1,422 @@
...
+        }
+        Ok(())
+    }
+}
+
+#[derive(Clone, Serialize, Deserialize, Default)]
+#[serde(default)]
+#[serde(rename_all = "kebab-case")]
+pub struct PdConfig {
+    pub endpoints: Vec<String>,
```
Comment (BusyJay) [inline src/config.rs]: "host:port" is an acceptable endpoint.

File: src/config.rs [hunk 4]
```diff
@@ -0,0 +1,422 @@
...
+            level0_slowdown_writes_trigger: 20,
+            level0_stop_writes_trigger: 36,
+            max_compaction_bytes: ReadableSize::gb(2),
+            compaction_pri: CompactionPriority::ByCompensatedSize,
+        }
+    }
+}
+
+impl CfConfig {
+    fn default_config() -> CfConfig {
```
Comment (siddontang) [inline src/config.rs]: should add a comment

**Comment by siddontang** [comment]
must update the etc/config-template.toml too

**Comment by siddontang** [comment]
PTAL @zhangjinpeng1987

**Comment by siddontang** [comment]
PTAL @andelf

**Comment by siddontang** [comment]
LGTM
Please fix conflict, PTAL @javaforfun

File: src/bin/tikv-server.rs [hunk 5]
```diff
@@ -936,17 +186,66 @@ fn run_raft_server(pd_client: RpcClient,
...
+    if let Some(data_dir) = matches.value_of("data-dir") {
+        config.storage.data_dir = data_dir.to_owned();
+    }
+
+    if let Some(endpoints) = matches.values_of("pd-endpoints") {
+        config.pd.endpoints = endpoints.map(|e| e.to_owned()).collect();
+    }
+
+    if let Some(labels_vec) = matches.values_of("labels") {
+        let mut labels = HashMap::new();
```
Comment (zhangjinpeng1987) [inline src/bin/tikv-server.rs]: HashMap can't guarantee the order of labels.

File: src/config.rs [hunk 6]
```diff
@@ -0,0 +1,418 @@
...
+use util::properties::{MvccPropertiesCollectorFactory, SizePropertiesCollectorFactory};
+use util::rocksdb::{FixedPrefixSliceTransform, FixedSuffixSliceTransform, NoopSliceTransform,
+                    CFOptions};
+
+const LOCKCF_MIN_MEM: usize = 256 * MB as usize;
+const LOCKCF_MAX_MEM: usize = GB as usize;
+const RAFTCF_MIN_MEM: usize = 256 * MB as usize;
+const RAFTCF_MAX_MEM: usize = 2 * GB as usize;
+
+fn memory_for_cf(cf: &str) -> usize {
```
Comment (zhangjinpeng1987) [inline src/config.rs]: memory_mb_for_cf ?

File: src/bin/tikv-server.rs (hunk 5 above)
Comment (BusyJay) [inline src/bin/tikv-server.rs]: Does the order of labels matter?

Comment (zhangjinpeng1987) [inline src/bin/tikv-server.rs]: Of course.

File: src/config.rs [hunk 7]
```diff
@@ -0,0 +1,418 @@
...
+            level0_slowdown_writes_trigger: 20,
+            level0_stop_writes_trigger: 36,
+            max_compaction_bytes: ReadableSize::gb(2),
+            compaction_pri: CompactionPriority::MinOverlappingRatio,
+        }
+    }
+}
+
+impl CfConfig {
+    fn write_config() -> CfConfig {
```
Comment (zhangjinpeng1987) [inline src/config.rs]: How to make association between `CfConfig` and `tikv.toml`?

Comment (BusyJay) [inline src/config.rs]: `toml::from_str` will convert content of tikv.toml to `TiKvConfig` automatically, which contains `CfConfig`.

File: src/bin/tikv-server.rs (hunk 5 above)
Comment (disksing) [inline src/bin/tikv-server.rs]: It's ok to be out of order. The order of pd's config is vital, not tikv's.

**Comment by BusyJay** [comment]
PTAL

**Comment by siddontang** [comment]
PTAL @javaforfun @zhangjinpeng1987

**Comment by siddontang** [comment]
@BusyJay 

Don't forget to update the document for the configuration.

**Comment by siddontang** [comment]
@LinuxGit  @liubo0127 

This PR will change configuration a lot, and it will break the compatibility. So for Ansible, we must use master branch to keep with the newest updates and supply a rc4, rc3, etc... branch for the released version.

**Comment by zhangjinpeng1987** [comment]
LGTM

**Comment by siddontang** [comment]
please fix the conflict. 

PTAL @javaforfun @overvenus

**Comment by javaforfun** [comment]
LGTM

**Pull request merged**